velox/exec/Spiller.cpp

```
@@ -0,0 +1,402 @@
+  // these will be materialized at a time for each stream of the
+  // merge.
+  constexpr int32_t kTargetBatchBytes = 1 << 20; // 1MB
+
+  RowVectorPtr spillVector;
+  auto& run = spillRuns_[partition];
+  try {
+    ensureSorted(run);
+    int64_t totalBytes = 0;
+    int32_t written = 0;
+    while (written < run.rows.size()) {
+      int32_t i = 0;
+      int32_t limit = std::min<uint64_t>(128, run.rows.size() - written);
+      int32_t bytes = 0;
+      for (; i < limit; ++i) {
+        bytes += container_.rowSize(run.rows[written + i]);
+        if (bytes > kTargetBatchBytes) {
+          ++i;
+          break;
+        }
+      }
+      folly::Range<char**> spilled(run.rows.data() + written, i);
+      extractSpill(spilled, &spillVector);
+      state_.appendToPartition(partition, spillVector);
+      written += i;
+      totalBytes += bytes;
+      if (totalBytes > maxBytes) {
+        break;
+      }
+    }
+    return std::make_unique<SpillStatus>(partition, written, nullptr);
+  } catch (const std::exception& e) {
+    return std::make_unique<SpillStatus>(
+        partition, 0, std::current_exception());
+  }
+}
+
+void Spiller::advanceSpill(uint64_t maxBytes) {
+  std::vector<std::shared_ptr<AsyncSource<SpillStatus>>> writes;
+  for (auto partition = 0; partition < spillRuns_.size(); ++partition) {
+    if (pendingSpillPartitions_.find(partition) ==
+        pendingSpillPartitions_.end()) {
+      continue;
+    }
+    writes.push_back(std::make_shared<AsyncSource<SpillStatus>>(
+        [partition, this, maxBytes]() {
+          return writeSpill(
+              partition, maxBytes / pendingSpillPartitions_.size());
+        }));
+    if (executor_) {
+      executor_->add([source = writes.back()]() { source->prepare(); });
+    }
+  }
+  auto sync = folly::makeGuard([&]() {
+    for (auto& write : writes) {
+      try {
+        write->move();
+      } catch (const std::exception& e) {
```

Would you comment why is it OK to ignore exceptions here?